### DISCUSSION TITLE
spaCy Project: Part-of-speech Tagging & Dependency Parsing packaging problem

### DISCUSSION THREAD

**Discussion by kanayer** [explosion/spaCy#9840] [Help: Coding & Implementations] [answered]
I used the project to train the spacy tagger and parser for Thai language on my custom dataset. Preprocessing, training and evaluation worked fine, but the packaging gives the following error:

> File "/usr/local/lib/python3.9/site-packages/catalogue/__init__.py", line 95, in get raise RegistryError(
> catalogue.RegistryError: Cant't find 'spacy.MultiHashEmbed.v1' in registry spacy -> architectures. Available names: spacy-legacy.CharacterEmbed.v1, spacy-legacy.HashEmbedCNN.v1, spacy-legacy.MaxoutWindowEncoder.v1, spacy-legacy.MishWindowEncoder.v1, spacy-legacy.MultiHashEmbed.v1, spacy-legacy.TextCatBOW.v1, spacy-legacy.TextCatCNN.v1, spacy-legacy.TextCatEnsemble.v1, spacy-legacy.Tok2Vec.v1, spacy-legacy.TransitionBasedParser.v1, spacy-transformers.Tok2VecTransformer.v1, spacy-transformers.TransformerListener.v1, spacy-transformers.TransformerModel.v1, spacy.CharacterEmbed.v2, spacy.EntityLinker.v1, spacy.HashEmbedCNN.v2, spacy.MaxoutWindowEncoder.v2, spacy.MishWindowEncoder.v2, spacy.MultiHashEmbed.v2, spacy.PretrainCharacters.v1, spacy.PretrainVectors.v1, spacy.SpanCategorizer.v1, spacy.Tagger.v1, spacy.TextCatBOW.v2, spacy.TextCatCNN.v2, spacy.TextCatEnsemble.v2, spacy.TextCatLowData.v1, spacy.Tok2Vec.v2, spacy.Tok2VecListener.v1, spacy.TorchBiLSTMEncoder.v1, spacy.TransitionBasedParser.v2
> 

Details about the OS and versions of spaCy and  Python:

```
spaCy version    3.1.2                         
OS               macOS 64bit
Python version   3.9.9  

```
             
Can you please help me understand what is wrong and how can I fix the issue?

**Comment by adrianeboyd** [comment, marked as answer]
Hi, this bug is fixed in v3.1.3.

**Reply by kanayer** [reply]
I have upgraded my spaCy from 3.1.2 to 3.1.3 and the issue still remains. When I try to run `spacy project run package `the same error appears:

>  File "/usr/local/lib/python3.9/site-packages/catalogue/__init__.py", line 95, in get raise RegistryError(
catalogue.RegistryError: Cant't find 'spacy.MultiHashEmbed.v1' in registry spacy -> architectures. Available names: spacy-legacy.CharacterEmbed.v1, spacy-legacy.HashEmbedCNN.v1, spacy-legacy.MaxoutWindowEncoder.v1, spacy-legacy.MishWindowEncoder.v1, spacy-legacy.MultiHashEmbed.v1, spacy-legacy.TextCatBOW.v1, spacy-legacy.TextCatCNN.v1, spacy-legacy.TextCatEnsemble.v1, spacy-legacy.Tok2Vec.v1, spacy-legacy.TransitionBasedParser.v1, spacy-transformers.Tok2VecTransformer.v1, spacy-transformers.TransformerListener.v1, spacy-transformers.TransformerModel.v1, spacy.CharacterEmbed.v2, spacy.EntityLinker.v1, spacy.HashEmbedCNN.v2, spacy.MaxoutWindowEncoder.v2, spacy.MishWindowEncoder.v2, spacy.MultiHashEmbed.v2, spacy.PretrainCharacters.v1, spacy.PretrainVectors.v1, spacy.SpanCategorizer.v1, spacy.Tagger.v1, spacy.TextCatBOW.v2, spacy.TextCatCNN.v2, spacy.TextCatEnsemble.v2, spacy.TextCatLowData.v1, spacy.Tok2Vec.v2, spacy.Tok2VecListener.v1, spacy.TorchBiLSTMEncoder.v1, spacy.TransitionBasedParser.v2

```
============================== Info about spaCy ==============================

spaCy version    3.1.3                         
Location         /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/spacy
Platform         macOS 64bit
Python version   3.9.9
```
The problem persists even after upgrading spaCy to version 3.2.1 
```

============================== Info about spaCy ==============================

spaCy version    3.2.1                         
Location         /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/spacy
Platform         macOS 64bit
Python version   3.9.9
```

**Reply by adrianeboyd** [reply]
Hmm, can you try explicitly upgrading `spacy-legacy` with `pip install -U spacy-legacy`?

**Reply by kanayer** [reply]
Thank you, it worked! One question, how to load the model after packaging it? I went into `dist `folder and did `pip install thai_ud-1.0.0` and it is installed into `/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages`. But how do I load my spacy model in IDE? 

I load spacy model the following way:
`nlp = spacy.load("/Users/kkk/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/th-ud-thai-1.0.0")
`
but get the following error:

> OSError: [E050] Can't find model '/Users/kkk/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/th-ud-thai-1.0.0'. It doesn't seem to be a Python package or a valid path to a data directory.

**Reply by adrianeboyd** [reply]
Once you've installed the package, you can load it with the package name that you see in `pip list`, so it's exactly the same as for `en_core_web_sm`:

```python
spacy.load("th_ud_thai")
```

For languages with additional requirements, you can add them to the package requirements in `meta.json` so they're installed automatically with the package like any other python dependency. You can pass in a partial `meta.json` to `spacy package --meta` that contains the additional requirements:

```json
{
  "requirements": ["pythainlp>=2.0"]
}
```

**Reply by kanayer** [reply]
But then I get an error 

> "ModuleNotFoundError: No module named 'th-ud-thai'"

Is there anything I am doing wrong?

**Reply by adrianeboyd** [reply]
Ah, I forgot that the `pip list` output doesn't convert the dashes into underscrores. Try: `spacy.load("th_ud_thai")`
